### ISSUE TITLE
Is there a way to attach properties to partial incremental paths?

### ISSUE THREAD

**Issue by NReilingh** [nightroman/Invoke-Build#143]
I am writing a partial incremental task that uses script blocks for inputs and outputs.

My inputs are based on "configuration data" that is an array of strings — the input script transforms this into a set of paths.

My outputs are also a set of paths based on this array of strings, but they are located in a different directory, so it would be more efficient to generate them based on the original string than based on the entire input path. So instead of using a process block when defining my outputs, I am using the same kind of pattern I use to define the inputs; running my original array through a foreach-object.

It would be nice to be able to pass more than just the input and output paths into the output and job process blocks. I made an attempt at this by adding a noteproperty to the path before returning it in the Inputs block, but this did not appear to work. Here's some example code for this pattern:

```powershell
# Synopsis: Unzip those files
task UnZipInstallPkg @{
    Partial = $true
    Inputs = {
        $VersionArray | ForEach-Object {
            Join-Path -Resolve $ZIPDIR "$_.zip" | Add-Member -NotePropertyName Version -NotePropertyValue $_ -PassThru
        }
    }
    Outputs = {
        process {
            Join-Path $OUTDIR $_.Version
        }
    }
    Jobs = 'CopyInstallZip', {
        begin {
            New-Item -ItemType "Directory" -Force -Path $OUTDIR | Out-Null
        }
        process {
            Expand-Archive -Path $_ -DestinationPath $2
        }
    }
}
```

Should this be possible, or should I try to do this a different way, or does this make for an interesting feature request?

**Comment by nightroman** [comment]
Is it what you are looking for?

```
. Invoke-Build
 help task -Parameter data
```

```
   -Data
       Any object attached to the task. It is not used by the engine.
       When the task is invoked this object is available as $Task.Data.
```

**Comment by NReilingh** [comment]
Hm, I'm not sure. As I read that, it looks like it's basically equivalent to declaring a variable outside of the task and then referencing it inside the task. (Perhaps more useful when calling other tasks programmatically from within another task?)

What I'm looking for is the ability to pass more than just a filename from Inputs through to the Process blocks of Outputs and the Job itself. That seems safer than what I'm doing now, which is to generate the inputs and outputs in two separate loops, which works, but doesn't have a guarantee that the arrays generated are going to have the same number of results, in the same order. If I generate my output files in a different order from the input files, and then use `$_` and `$2` in the job, I may end up with swapped outputs.

**Comment by nightroman** [comment]
Ah, I see, sorry for the first too quick suggestion.

I think in your use case `Outputs` should not use the `process` block at all.
It should generate paths from the same data and in the similar way as `Inputs`:

```powershell
task UnZipInstallPkg @{
    Partial = $true
    Inputs = {
        $VersionArray | ForEach-Object {
            Join-Path -Resolve $ZIPDIR "$_.zip"
        }
    }
    Outputs = {
        $VersionArray | ForEach-Object {
            Join-Path $OUTDIR $_
        }
    }
    Jobs = {
        begin {
            New-Item -ItemType "Directory" -Force -Path $OUTDIR | Out-Null
        }
        process {
            # just for testing
            "processing $_ -> $2"
            Copy-Item $_ $2
        }
    }
}
```

**Comment by nightroman** [comment]
Alternatively, `Inputs` may generate both input and output paths with any
additional data attached and store such objects in the script variable.
Then this variable is used in both `Inputs` and `Outputs` blocks:

```powershell
    Inputs = {
        $script:MyInOutData = $VersionArray | ForEach-Object {
            @{
                in = Join-Path -Resolve $ZIPDIR "$_.zip"
                out = Join-Path $OUTDIR $_
            }
        }
        $script:MyInOutData | ForEach-Object { $_.in }
    }
    Outputs = {
        $script:MyInOutData | ForEach-Object { $_.out }
    }
```

**Comment by nightroman** [comment]
I have a different question/issue about your scenario. Your output paths seem to be directories, not files, right? Invoke-Build expects files, I think. So that your outputs are always treated as missing = out-of-date. Or am I missing something?

**Comment by nightroman** [comment]
Added the tip and reference to this topic to wiki, see the end of
https://github.com/nightroman/Invoke-Build/wiki/Partial-Incremental-Tasks

**Comment by NReilingh** [comment]
Aha, you're right about the directory thing — that's just an oversight in the example I provided. Usually what I do is use a zero-length file inside the directory as my output, and New-Item that file at the end of the Job process block.

Thank you for the tip about using `$Task.Data` as a task-scoped persistence layer! Works great.

Here's the pattern I ended up going with, which may err on the side of being overly complex, but satisfies my compulsion to avoid having separate loops generate paths that need a 1:1 correspondence. Since the paths defined in Inputs are always passed through to Outputs and Jobs, I'm using those paths to do a key lookup in `$Task.Data` as a hashtable.

```powershell
Inputs = {
    $Task.Data = @{}
    $VersionArray | ForEach-Object {
        $Task.Data.Add((Join-Path -Resolve $INDIR "$_.zip"), $_)
    }
    $Task.Data.Keys
}
Outputs = {
    process {
        Join-Path $OUTDIR "$($Task.Data[$_]).zip"
    }
}
Jobs = {
    process {
        Copy-Item $_ $2
    }
}
```
